### PR TITLE
feat(product): Add 'product' commands for each Fastly product.

### PR DESCRIPTION
**Pull request by kpfleming**
Adds 'enable', 'disable', and 'status' commands for each product previously supported by the 'products' command (which is now deprecated).

Also adds 'configure' commands for products which support configuration (DDoS Protection and Next-Gen WAF).

### CODE REVIEW

File: pkg/commands/product/bot_management/status.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	if s.Enabled {
+		msg = "enabled"
+	} else {
+		msg = "disabled"
+	}
+
+	text.Info(out,
+		bot_management.ProductName+" is %s on service %s", msg, serviceID)
+
+	return nil
```
Comment (cee-dub) [inline pkg/commands/product/bot_management/status.go:111]: This logic is a bit hard to follow. One idea...
```suggestion
        state := "disabled"
        _, err = GetFn(c.Globals.APIClient, serviceID)
	if err != nil {
                // The API returns a 'Bad Request' error when the product has
		// not been enabled on the service so we ignore it; any other
		// error should be reported
                var herr *fastly.HTTPError
		if !errors.As(err, &herr) || !herr.IsBadRequest() {
			c.Globals.ErrLog.Add(err)
			return err
		}
        } else {
                s.Enabled = true
                state = "enabled"
        }

        if ok, err := c.WriteJSON(out, s); ok {
		return err
	}

        text.Info(out, bot_management.ProductName+" is %s on service %s", state, serviceID)
	return nil
```

File: pkg/commands/product/bot_management/status.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
+package bot_management
```
Comment (cee-dub) [inline pkg/commands/product/bot_management/status.go:1]: The linter is agreeing with me that package names shouldn't have underscores.

File: pkg/commands/product/bot_management/product_test.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+			Name:      "validate invalid json/verbose flag combo: status",
+			Args:      "status --service-id 123 --json --verbose",
+			WantError: "invalid flag combination, --verbose and --json",
+		},
+		{
+			Name: "validate success for enabling product",
+			Setup: func(t *testing.T, scenario *testutil.CLIScenario, opts *global.Data) {
+				sub.EnableFn = func(_ api.Interface, _ string) (*bot_management.EnableOutput, error) {
+					return nil, nil
+				}
```
Comment (cee-dub) [inline pkg/commands/product/bot_management/product_test.go:43]: Cool!

File: pkg/commands/product/bot_management/disable.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+// NewDisableCommand returns a usable command registered under the parent.
+func NewDisableCommand(parent argparser.Registerer, g *global.Data) *DisableCommand {
+	c := DisableCommand{
+		Base: argparser.Base{
+			Globals: g,
+		},
+	}
+	c.CmdClause = parent.Command("disable", "Disable the "+bot_management.ProductName+" product")
```
Comment (cee-dub) [inline pkg/commands/product/bot_management/disable.go:37]: At first, I was rather confused until I realized that `bot_management.ProductName` refers to `github.com/fastly/go-fastly/v9/fastly/products/bot_management.ProductName` and not to something in this package. Same for the rest of the `bot_management.Thing`s in `pkg/commands/product/bot_management/*.go` files.

**Review by cee-dub — COMMENTED** [review]
Love the test mocking. Package names could be more idiomatic.

File: pkg/commands/product/bot_management/disable.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/fastly/cli/pkg/global"
+	"github.com/fastly/cli/pkg/manifest"
+	"github.com/fastly/cli/pkg/text"
+)
+
+// DisableFn is a dependency-injection point for unit tests to provide
+// a mock implementation of the API operation.
+var DisableFn = func(client api.Interface, serviceID string) error {
+	return bot_management.Disable(client.(*fastly.Client), serviceID)
+}
```
[the remark this answers is not in the document]
Comment (kpfleming) [reply pkg/commands/product/bot_management/disable.go:20]: Yeah, it's fairly hard to understand.

File: pkg/commands/product/bot_management/status.go (hunk 1 above)
Comment (kpfleming) [reply pkg/commands/product/bot_management/status.go:111]: Ahh, yes, that's definitely better. I'll roll this into some other changes in the branch.

File: pkg/commands/product/bot_management/status.go (hunk 2 above)
Comment (kpfleming) [reply pkg/commands/product/bot_management/status.go:1]: Yeah... I've been trying to stick with using our actual product IDs for ease of understanding by future maintainers, but the Go community has a different opinion (except for test packages, where underscores are allowed).

File: pkg/commands/product/bot_management/disable.go (hunk 4 above)
Comment (kpfleming) [reply pkg/commands/product/bot_management/disable.go:37]: I was confused by that when I started working in Go; if you're in package A, and you've also imported a package named A, then qualified references like `A.foo` *always* refer to the imported package and never to the 'local' package. Now that I've gotten used to it I'm not bothered by it, but this sort of thing is what prompted the desire for import aliases in the other project :-)

File: pkg/commands/product/bot_management/status.go (hunk 1 above)
Comment (cee-dub) [reply pkg/commands/product/bot_management/status.go:111]: Sorry, I didn't use my normal editor and looks like I have a syntax error in there.

File: pkg/commands/product/bot_management/status.go (hunk 2 above)
Comment (cee-dub) [reply pkg/commands/product/bot_management/status.go:1]: In that case it is representing a build tag. The same works for other specific words that match OS targets or architectures, like `_linux` or `_arm64`.

Comment (cee-dub) [reply pkg/commands/product/bot_management/status.go:1]: There also the only exception to two packages in one directory, where `p` and `p_test` can coexist to allow for blackbox testing of `p`.

File: pkg/commands/product/bot_management/disable.go (hunk 4 above)
Comment (cee-dub) [reply pkg/commands/product/bot_management/disable.go:37]: Might I suggest an alias is probably appropriate here? `bot` or `botapi` could work

File: pkg/commands/product/bot_management/disable.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/fastly/cli/pkg/api"
+	"github.com/fastly/cli/pkg/argparser"
+	"github.com/fastly/cli/pkg/global"
+	"github.com/fastly/cli/pkg/manifest"
+	"github.com/fastly/cli/pkg/text"
+)
+
+// DisableFn is a dependency-injection point for unit tests to provide
+// a mock implementation of the API operation.
+var DisableFn = func(client api.Interface, serviceID string) error {
```
Comment (cee-dub) [inline pkg/commands/product/bot_management/disable.go:18]: Probably a Go idiom or just a style preference of my own from familiarity with the standard library, but i suggest `DisableFunc` and friends.

File: pkg/commands/product/bot_management/disable.go (hunk 4 above)
Comment (cee-dub) [reply pkg/commands/product/bot_management/disable.go:37]: Try it and see what feels best, I guess!

File: pkg/commands/product/bot_management/disable.go (hunk 6 above)
Comment (kpfleming) [reply pkg/commands/product/bot_management/disable.go]: Resolved as part of some refactoring.

**Comment by kpfleming** [comment]
@cee-dub If you're interested, this could use a second 'structural' review; much like go-fastly, this now has an `internal/productcore` package which uses generics to provide the base implementations for all of the commands in `fastly product <name> ...`. It's complete for all products except NGWAF, which requires an extra enablement parameter, and then it will get extended to support the configuration commands too.

File: pkg/commands/product/botmanagement/common.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+	"github.com/fastly/go-fastly/v9/fastly"
+	"github.com/fastly/go-fastly/v9/fastly/products/botmanagement"
+
+	"github.com/fastly/cli/internal/productcore"
+	"github.com/fastly/cli/pkg/api"
+)
+
+// EnablementHooks is a structure of dependency-injection points used
+// by unit tests to provide mock behaviors
+var EnablementHooks = productcore.EnablementHookFuncs[*botmanagement.EnableOutput]{
```
Comment (cee-dub) [inline pkg/commands/product/botmanagement/common.go:13]: The `*botmanagement.EnableOutput` here confused me at first, and I was wondering why there were all these different types when everything here is nicely generic, but then I realized it is the `go-fastly` API response. Not certain this calls for an import alias, but I would find it reasonable to use one here.

File: internal/productcore/disable.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+	"github.com/fastly/cli/pkg/argparser"
+	"github.com/fastly/cli/pkg/global"
+	"github.com/fastly/cli/pkg/text"
+)
+
+// Disable is a base type for all 'disable' commands.
+type Disable[O any] struct {
+	Base
+	hooks *EnablementHookFuncs[O]
```
Comment (cee-dub) [inline internal/productcore/disable.go:16]: I'm not sure of any reason that `hooks` needs to be of pointer type. Should be fine to use the struct's value here and in the other generic action implementations.

File: pkg/commands/product/botmanagement/common.go [hunk 9]
```diff
@@ -0,0 +1,23 @@
+package botmanagement
```
Comment (cee-dub) [inline pkg/commands/product/botmanagement/common.go:1]: I think this package and each of the other new enable/disable/status packages under `pkg/commands/product` should be reduced to fewer files. When they each contained full implementations, separate files made sense. Now that the generic core does most of the work, I'd prefer to see something like `commands.go` and `commands_test.go` until a product gains unique functionality.

**Review by cee-dub — COMMENTED** [review]
Nice use of generics to deal with consuming the different API type that are essentially the same.

File: go.mod [hunk 10]
```diff
@@ -78,3 +79,5 @@ require (
 )
 
 require 4d63.com/optional v0.2.0
+
+replace github.com/fastly/go-fastly/v9 => github.com/kpfleming/go-fastly/v9 v9.12.1-0.20250113130142-b6a3af55bc44
```
[the remark this answers is not in the document]
Comment (kpfleming) [reply go.mod:83]: https://github.com/fastly/go-fastly/pull/572

That will be released before this is merged.

File: pkg/commands/product/botmanagement/common.go (hunk 7 above)
Comment (kpfleming) [reply pkg/commands/product/botmanagement/common.go:13]: Yeah... I thought about using an import alias here, as I had to do search-and-replace in 9 sets of source files :-)

File: internal/productcore/disable.go (hunk 8 above)
Comment (kpfleming) [reply internal/productcore/disable.go:16]: During testing the 'global' EnablementHookFuncs structure in each package gets modified by the tests, and so if a copy was made during the instantiation of this structure (which happens during the CLI startup) the mock functions would not be called.

File: pkg/commands/product/botmanagement/common.go (hunk 9 above)
Comment (kpfleming) [reply pkg/commands/product/botmanagement/common.go:1]: Very good point... I'll do that now, and also use import aliases to remove unnecessary differences between the source files.

File: internal/productcore/disable.go (hunk 8 above)
Comment (kpfleming) [reply internal/productcore/disable.go:16]: I'll add a comment to the source during the next round of edits, so the next person with this question will have an answer!
